### PR TITLE
feat: add basic auth authentication for securing grpc

### PR DESCRIPTION
**Pull request by Ja7ad**
Implementation of: https://pips.pactus.org/PIPs/pip-20

### CODE REVIEW

**Comment by Ja7ad** [comment]
@b00f @kehiy @amirvalhalla I fix conflict, please review it.

File: util/htpasswd/htpasswd.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+// CompareBasicAuth compare basic auth with bcrypt algorithm.
+func CompareBasicAuth(basicAuthCredential, user, password string) error {
+	parsedUser, parsedHashedPass, err := ParseHtpasswdAuth(basicAuthCredential)
+	if err != nil {
+		return err
+	}
+
+	if parsedUser != user {
+		return errors.New("user is invalid")
```
Comment (b00f) [inline util/htpasswd/htpasswd.go:25]: Define errors separately.

File: util/htpasswd/htpasswd_test.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+			user:        "user",
+			password:    "wrongpassword",
+			expectError: true,
+		},
+		{
+			name:        "MalformedCredential",
+			input:       "malformed",
+			user:        "user",
+			password:    "foobar",
+			expectError: true,
```
Comment (b00f) [inline util/htpasswd/htpasswd_test.go:86]: Specify which error is expected. If it is `nil`, then it should pass without error. Specifying the error type helps ensure that the proper error is captured.

File: util/htpasswd/htpasswd_test.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+			wantPass: "",
+			wantErr:  true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			user, pass, err := ExtractBasicAuthFromContext(tt.ctx)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("ExtractBasicAuthFromContext() error = %v, wantErr %v", err, tt.wantErr)
```
Comment (b00f) [inline util/htpasswd/htpasswd_test.go:208]: Please use `testify` package and check the error type

File: util/htpasswd/htpasswd.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	}
+
+	return bcrypt.CompareHashAndPassword([]byte(parsedHashedPass), []byte(password))
+}
+
+// ParseHtpasswdAuth parse htpasswd auth.
+func ParseHtpasswdAuth(auth string) (user, encodedPassword string, err error) {
```
Comment (b00f) [inline util/htpasswd/htpasswd.go:32]: No name for return values

File: util/htpasswd/htpasswd.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	encodedPassword = parts[1]
+
+	return
+}
+
+// ExtractBasicAuthFromContext extract basic auth from incoming context in grpc request.
+func ExtractBasicAuthFromContext(ctx context.Context) (user, password string, err error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return "", "", errors.New("metadata not found")
```
Comment (b00f) [inline util/htpasswd/htpasswd.go:48]: Same here, please define errors

File: www/grpc/config.go [hunk 6]
```diff
@@ -1,10 +1,11 @@
...
 
 type Config struct {
-	Enable       bool          `toml:"enable"`
-	EnableWallet bool          `toml:"enable_wallet"`
-	Listen       string        `toml:"listen"`
-	Gateway      GatewayConfig `toml:"gateway"`
+	Enable              bool          `toml:"enable"`
+	EnableWallet        bool          `toml:"enable_wallet"`
+	Listen              string        `toml:"listen"`
+	BasicAuthCredential string        `toml:"basic_auth_credential"`
```
Comment (b00f) [inline www/grpc/config.go:7]: I think we can simply use "Auth" for this config

File: config/example_config.toml [hunk 7]
```diff
@@ -153,6 +153,13 @@
   # `listen` is the address to listen for incoming connections for gRPC server.
  ## listen = "127.0.0.1:50051"
 
+  # `basic_auth_credential` is basic auth for authentication layer for improve security of gRPC (https://pips.pactus.org/PIPs/pip-20),
```
Comment (b00f) [inline config/example_config.toml:156]: Please keep the comment shorter without a link. No need to mention PIP.
Note that we will have a "How to generate a password hash for gRPC?" article on the website soon.

File: config/example_config.toml [hunk 8]
```diff
@@ -153,6 +153,13 @@
...
+  # you can generate user and password (hashed) via htpasswd util apache
+  # you can check util via https://httpd.apache.org/docs/2.4/programs/htpasswd.html
+  # for generate use command htpasswd -bnBC 10 {user} {password} and
```
Comment (b00f) [inline config/example_config.toml:159]: We may have different methods to generate password hash.

File: www/grpc/middleware_test.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+			name:          "NoMetadata",
+			authHeader:    "",
+			expectedError: codes.Unauthenticated,
+		},
+		{
+			name:          "MalformedAuthHeader",
+			authHeader:    malformedAuth,
+			expectedError: codes.Unauthenticated,
+		},
+		// More test cases if necessary...
```
Comment (b00f) [inline www/grpc/middleware_test.go]: Remove this line

File: www/grpc/middleware_test.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+				md := metadata.New(map[string]string{"authorization": tt.authHeader})
+				ctx = metadata.NewIncomingContext(ctx, md)
+			}
+
+			interceptor := BasicAuth("user:$2y$10$5Kjd955BDWLouqckHzBjKuCF6hFOUD61lhm8QpjDVHTUwMIrYUdq2")
+
+			_, err := interceptor(ctx, nil, &grpc.UnaryServerInfo{}, mockUnaryHandler)
+
+			if got, want := status.Code(err), tt.expectedError; got != want {
+				t.Errorf("expected error code %v, got %v", want, got)
```
Comment (b00f) [inline www/grpc/middleware_test.go:64]: Make sure we use `testify`  package to assert errors

**Review by b00f — APPROVED** [review]

File: config/example_config.toml (hunk 8 above)
Comment (Ja7ad) [reply config/example_config.toml:159]: Sure, but we suggestion htpasswd for generate basic auth.

File: config/example_config.toml (hunk 7 above)
Comment (Ja7ad) [reply config/example_config.toml:156]: Ok

File: www/grpc/config.go (hunk 6 above)
Comment (Ja7ad) [reply www/grpc/config.go:7]: Auth have any method (token base, basic auth, api key and etc...) and we can't use for a method this name, this is not explicit name.

File: util/htpasswd/htpasswd_test.go [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+			input:           "user3:$2y$10$eKLWzld7iMPrcyDqam8.Y.R1deeSUBWFD3P6eQHJ0Iqa1qR4yBxaq",
+			user:            "user3",
+			encodedPassword: "$2y$10$eKLWzld7iMPrcyDqam8.Y.R1deeSUBWFD3P6eQHJ0Iqa1qR4yBxaq",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.user, func(t *testing.T) {
+			user, encodedPass, err := ParseHtpasswdAuth(tt.input)
+			if err != nil {
```
Comment (b00f) [inline util/htpasswd/htpasswd_test.go:43]: It can be replaced by: `require.NoError(t, err)`

**Review by b00f — APPROVED** [review]

**Review by b00f — APPROVED** [review]

**Comment by b00f** [comment]
@Ja7ad merged with thanks
Now, we can add a post on website to explain: "How to secure gRPC using basic authentication?"

**Pull request merged**